torch/nn/utils/parametrize.py

```diff
@@ -300,17 +300,31 @@ def _inject_property(module: Module, tensor_name: str) -> None:
     # This should never fire if register_parametrization is correctly implemented
     assert not hasattr(module, tensor_name)
 
-    def get_parametrized(self) -> Tensor:
+    @torch.jit.unused
+    def get_cached_parametrization(parametrization) -> Tensor:
         global _cache
+        key = (id(module), tensor_name)
+        tensor = _cache.get(key)
+        if tensor is None:
+            tensor = parametrization()
+            _cache[key] = tensor
+        return tensor
 
+    def get_parametrized(self) -> Tensor:
         parametrization = self.parametrizations[tensor_name]
         if _cache_enabled:
-            key = (id(module), tensor_name)
-            tensor = _cache.get(key)
-            if tensor is None:
-                tensor = parametrization()
-                _cache[key] = tensor
-            return tensor
+            if torch.jit.is_scripting():
+                # Scripting
+                raise RuntimeError('Caching is not implemented for scripting. '
+                                   'Either disable caching or avoid scripting.')
+            elif torch._C._get_tracing_state() is not None:
+                # Tracing
+                raise RuntimeError('Caching should be disabled while tracing a '
+                                   'model. You can run a traced model with '
```

Comment:
The caching flag will actually be ignored while running the traced model no?
If so, I think this message is a bit misleading as I feel like it implies that the flag will be honored when running the traced model.

Reply:
The caching flag might be enabled during the tracing. This line is here to catch these types of implementations::

```python
with parametrize.cached():
    traced_model = torch.jit.trace_module(model, {'forward': x})
```

I think the message could be improved. How about

```suggestion
                                   'Cannot trace a model while caching parametrizations.'
```

Alternatively, we don't throw an error here, and quietly just do 

```suggestion
                                   return parametrization()
```

Reply:
No, the error is good. And the updated error message is good as well I think!